frontend/src/composables/use-analytics.ts

```diff
@@ -0,0 +1,162 @@
+ * client-side. This excludes props that need `window`.
+ *
+ * @param $ua - the User-Agent information from `express-useragent`
+ * @param i18n - the Nuxt i18n instance from `@nuxtjs/i18n`
+ */
+export const getIsomorphicProps = ({ $ua, i18n }: Context | ContextReturn) => ({
+  timestamp: new Date().toISOString(),
+  language: i18n.locale,
+  ...($ua
+    ? {
+        ua: $ua.source,
+        os: $ua.os,
+        platform: $ua.platform,
+        browser: $ua.browser,
+        version: $ua.version,
+      }
+    : {}),
+})
+
+/**
+ * Get Plausible props that work only on the client-side. This only includes
+ * props that need `window`.
+ */
+export const getWindowProps = () => ({
+  origin: window.location.origin,
+  pathname: window.location.pathname,
+  referrer: window.document.referrer,
+  width: window.innerWidth,
+  height: window.innerHeight,
+})
+
+/**
+ * Get Plausible props present in `window` through alternate channels. This uses
+ * Nuxt context fields that are only present on the server-side.
+ *
+ * @param req - the request received by the Nuxt SSR server
+ * @param route - the Vue router route being rendered from `vue-router`
+ * @param $cookies - the cookies object from `cookie-universal-nuxt`
+ */
+export const getPseudoWindowProps = ({
+  req,
+  route,
+  $cookies,
+}: Context | ContextReturn) => ({
+  origin: req.headers.host ?? "", // `Host` header includes port.
+  pathname: isRef(route) ? route.value.fullPath : route.fullPath,
+  referrer: req.headers.referer ?? "",
+  width: $cookies.get("uiDeviceWidth") ?? -1,
+  height: $cookies.get("uiDeviceHeight") ?? -1,
+})
+
+/**
+ * The `ctx` parameter must be supplied if using this composable outside the
+ * bounds of the composition API.
+ *
+ * @param ctx - the Nuxt context
+ */
+export const useAnalytics = (ctx?: Context) => {
+  const context = ctx ?? useContext()
+  const { $cookies, $plausible, route, req } = context
+
+  /**
+   * Send a custom event to Plausible. Mandatory props are automatically merged
+   * with the event-specific props.
+   *
+   * @param name - the name of the event being recorded
+   * @param payload - the additional information to record about the event
+   */
+  const sendCustomEvent = <T extends EventName>(
+    name: T,
+    payload: Events[T]
+  ) => {
+    $plausible.trackEvent(name, {
+      props: {
+        ...getIsomorphicProps(context),
+        ...getWindowProps(),
+        ...payload, // can override mandatory props
+      },
+    })
+  }
+
+  /**
+   * Send a custom event to Plausible. Mandatory props are automatically merged
+   * with the event-specific props. This function is meant to be used on the
+   * server side because it uses their API instead of using the integration.
+   *
+   * @param name - the name of the event being recorded
+   * @param payload - the additional information to record about the event
+   */
+  const sendCustomEventApi = async <T extends EventName>(
```

Comment:
So I could be totally wrong about this, and please let me know if I'm missing something big here. I really think we might be able to significantly simplify this implementation. I think we're doing a lot of duplicated work here by proxying the API call in this particular way. The recommended proxy approach is to have something like Nginx just forward the call from the script to plausible and let the script handle the rest of the parameters. See the example here: https://plausible.io/docs/proxy/guides/nginx

In that case, we would still be using our own domain for the events, but we don't have to handle _anything_ near the level of complexity in this function because we know that Plausible is passing the event through anyway. We would still have the flexibility to configure Nginx locally to send events to the docker-compose plausible instance rather than the production instance.

There is the possibility that some events would be fired on the server-side as well, like the `INSUFFICIENT_RESULTS`, unless we intentionally defer it to happen on the client side. If we want to defer it to happen on the client side, we need a mechanism to do that (a store? I don't know). However, I think this event could be implemented in the Django API instead and we could still track the event and even forward the information Plausible uses to calculate the user identifier from there.

Looking at the list of events from the events RFC, the only events that have this quality are the server rendered event and the insufficient results one. The rest all do come as a result of user interaction with the page and would never happen during the server rendering of a page.

If we can eliminate those events from the frontend code, then we pretty significantly simplify this stuff because `sendCustomEventApi` and `sendCustomEvent` can (a) ignore the difficulties of isomorphism, (b) people writing code that fires events don't need to choose from two different similarly named functions, (c) we eliminate the need to maintain a lot of this code and would no longer need the server middleware to forward events (we'd rely on Nginx to forward them for us instead).

Additionally, I think even the server rendered event might be redundant with the [entry pages](https://plausible.io/docs/metrics-definitions#entry-pages) metric. Any entry page (and their rates) would show is the number of server rendered pages (though requests already do that). Looking at that metrics page, I think we're duplicating a lot of stuff potentially. For example, do we need to gather screen dimensions if [Plausible already tracks it](https://plausible.io/docs/metrics-definitions#screen-size) (though it only tracks width, not height).

Here's basically the pit-fall that comes from not having considered the particular provider ahead of time. Plausible appears to have a great deal of things built into that we'd be making redundant with some of the stuff we're writing into the application here (all of which comes with added complexity). It made sense to assume that none of this stuff would be tracked when we weren't using a specific analytics provider, but now that we know exactly who we're using, we can rely on their default behaviours. I assume that if we were to switch providers it would be a significant enough deal that we'd need to do more work anyway. Trying to hand-spin this stuff on our own to not feel as locked-in might be a red herring.

Anyway, my recommendation here boils down to:

1. Use the regular Nginx approach documented by Plausible to proxy the script and the events API on our own domain.
2. Use the default Plausible window function to send custom events. If we need a wrapper to handle some details automatically then :+1:, but it can call the regular Plausible script function, and we'd avoid additional complexity.
3. Send the "insufficient results" custom event from Django.
4. Confirm that Plausible doesn't already track things for us that we're tracking. Entry page as an alias for "server rendered page" is a good example of this. Screen width as well (do we absolutely _need_ height?).

Overall, I think this would significant reduce the complexity of this work.

Reply:
This was an experiment, in public view, to see the boundaries of our event sending system, it was not supposed to go into production as-is (hence the draft PR).

In the end, the implementation of analytics is as follows.

- `vue-plausible` bundles the Plausible tracker into the code, removing the need to include via a script. This allows us to not be blocked by mechanisms like adblockers or DNS denylists because the tracker is shipped with the app code itself.

- `@nuxtjs/proxy` module is used to send all Plausible calls to the same domain as the app, so those calls will be made to `https://openverse.org/api/event`. The SSR server (Express afaik) will forward them to `https://plausible.io/api/event` (or wherever we host our Plausible instance), again bypassing adblockers and DNS denylists.

- All events will be sent from the client side, for the most part, so they will have access to `window`. The cookie code has been removed 7e06869. It was a test to see if a server, given enough info in cookies can emulate analytics sent from clients, not production ready and to be dropped before undrafting.

- Plausible tracks a few things that we don't need, like IP addresses which it uses to identify unique users. Considering that, like you said, I have not included IP in our custom mandatory props. Plausible also tracks screen width, like you said, but does not make it available to our custom goals (only using it to determine device afaik) so we must track that (and height for added measure) on our own. Plausible does not charge for data contained per event so we can make it as rich as we like.

- Server side events like `INSUFFICIENT_RESULTS` can be considered in a later phase of this project after this "MVP". We have plenty of learnings to derive from other events before we get to a point where lack of events data will prevent us from drawing conclusions.